### PR TITLE
Enhance matmul_op to support 4-D inputs

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/7319

### CODE REVIEW

File: paddle/operators/math/matmul.h [hunk 1]
```diff
@@ -41,10 +41,23 @@ class MatMulFunctor {
...
-    PADDLE_ENFORCE_LE(dim_b.size(), 3,
-                      "Input tensor b must be at most 3-dimensional.");
+
+    std::vector<int64_t> out_dim;
+    int64_t batch_count = 1;
+    if (dim_a.size() > 3) {
+      PADDLE_ENFORCE(dim_b.size() > 3,
+                     "The dimensions of X and Y must be the same, and both of "
+                     "them should be %d-dimensional.",
+                     dim_b.size());
```
Comment (lcy-seso) [inline paddle/operators/math/matmul.h]: It seems that the check in line48 is not consistent with its comments. It seems that the check should be like this:

```cpp
PADDLE_ENFORCE_EQ(dim_a.size(), dim_b.size(),
                  "The dimensions of X and Y must be the same."
                  "X_dim vs. Y_dim = %d vs. %d",
                  dim_a.size(), dim_b.size());
```

File: paddle/operators/math/matmul.h [hunk 2]
```diff
@@ -41,10 +41,23 @@ class MatMulFunctor {
...
+      // The previous Rank-2 dimensions are accumulated on the batch_count.
```
Comment (lcy-seso) [inline paddle/operators/math/matmul.h]: Does "previous Rank-2" here mean the first two dimensions?

File: paddle/operators/math/matmul.h [hunk 3]
```diff
@@ -41,10 +41,23 @@ class MatMulFunctor {
...
+      for (int j = 0; j < dim_a.size() - 2; ++j) {
+        PADDLE_ENFORCE(dim_b[j] == dim_a[j],
+                       "The dimensions of X[%d] and Y[%d] must be the same.", j,
+                       j);
```
Comment (lcy-seso) [inline paddle/operators/math/matmul.h]: - `PADDLE_ENFORCE` --> `PADDLE_ENFORCE_EQ`
- the comments:
```cpp
PADDLE_ENFORCE_EQ(dim_b[j], dim_a[j],
                   "The %d-th dimension of X[%d] and Y[%d] must be the same.", j);
```

File: paddle/operators/matmul_op.cc [hunk 4]
```diff
@@ -162,10 +187,14 @@ Examples without transpose:
 - X: [B, M, K], Y: [K] => Out: [B, M]
 - X: [M, K], Y: [B, K, N] => Out: [B, M, N]
 - X: [B, M, K], Y: [B, K, N] => Out: [B, M, N]
+- X: [B, ..., M, K], Y: [B, ..., K, N] => Out: [B, ..., M, N]
 
 The behavior is designed to be similar to the `numpy.matmul` function.
 The differences are:
-- Currently only rank 1 to rank 3 input tensors are supported.
+- When the rank of the input is greater than 3, the rank of X and
+  Y must be equal, and the former rank-2 dimensions are equal.
```
Comment (lcy-seso) [inline paddle/operators/matmul_op.cc]: former rank-2 dimensions --> the first two dimentsions

File: paddle/operators/matmul_op.h [hunk 5]
```diff
@@ -137,6 +137,12 @@ class MatMulGradKernel : public framework::OpKernel<T> {
       y_dims.push_back(1);
     }
 
+    int batch_count = 0;
+    // The previous Rank-2 dimensions are accumulated on the batch_count.
```
Comment (lcy-seso) [inline paddle/operators/matmul_op.h]: previous Rank-2  --> the first 2 dimenstions.

File: paddle/operators/math/matmul.h (hunk 2 above)
Comment (lcy-seso) [reply paddle/operators/math/matmul.h]: Yes, I see. Thank you.

File: paddle/operators/matmul_op.cc [hunk 6]
```diff
@@ -162,10 +189,14 @@ Examples without transpose:
...
 - X: [B, M, K], Y: [B, K, N] => Out: [B, M, N]
+- X: [B, ..., M, K], Y: [B, ..., K, N] => Out: [B, ..., M, N]
 
 The behavior is designed to be similar to the `numpy.matmul` function.
 The differences are:
-- Currently only rank 1 to rank 3 input tensors are supported.
+- When the rank of the input data is less than or equal to 3, it
+  is similar to the `numpy.matmul` function.
+- When the rank of the input is greater than 3, the rank of X and
+  Y must be equal, and the front `rank - 2` dimensions must be equal.
```
Comment (lcy-seso) [inline paddle/operators/matmul_op.cc]: front here should be first.

File: paddle/operators/math/matmul.h [hunk 7]
```diff
@@ -41,10 +41,24 @@ class MatMulFunctor {
...
-                      "Input tensor b must be at most 3-dimensional.");
+
+    std::vector<int64_t> out_dim;
+    int64_t batch_count = 1;
+    if (dim_a.size() > 3) {
+      PADDLE_ENFORCE(dim_b.size() == dim_a.size(),
+                     "The dimensions of X and Y must be the same, and both of "
+                     "them should be %d-dimensional.",
+                     dim_b.size());
+      // The front rank-2 dimensions are accumulated on the batch_count, and the
```
Comment (lcy-seso) [inline paddle/operators/math/matmul.h]: the `front` here should be `first`

File: paddle/operators/matmul_op.cc [hunk 8]
```diff
@@ -41,10 +41,26 @@ class MatMulOp : public framework::OperatorWithKernel {
...
+    std::vector<int64_t> out_dim;
+    int64_t batch_count = 1;
+    if (dim_x.size() > 3) {
+      PADDLE_ENFORCE_EQ(
+          dim_y.size(), dim_x.size(),
+          "The dimensions of X and Y must be the same, and both of "
+          "them should be %d-dimensional.",
+          dim_x.size());
+
+      // The front rank-2 dimensions are accumulated on the batch_count, and the
```
Comment (lcy-seso) [inline paddle/operators/matmul_op.cc]: the front here should be first.

File: paddle/operators/matmul_op.h [hunk 9]
```diff
@@ -137,6 +137,13 @@ class MatMulGradKernel : public framework::OpKernel<T> {
       y_dims.push_back(1);
     }
 
+    int batch_count = 0;
+    // The front rank-2 dimensions are accumulated on the batch_count, and the
```
Comment (lcy-seso) [inline paddle/operators/matmul_op.h]: the `front` here should be `first`

File: paddle/operators/math/matmul.h (hunk 7 above)
Comment (chengduoZH) [reply paddle/operators/math/matmul.h]: Done. Many thanks!

File: paddle/operators/matmul_op.cc (hunk 8 above)
Comment (chengduoZH) [reply paddle/operators/matmul_op.cc]: Done

File: paddle/operators/matmul_op.cc (hunk 6 above)
Comment (chengduoZH) [reply paddle/operators/matmul_op.cc]: Done

File: paddle/operators/matmul_op.h (hunk 9 above)
Comment (chengduoZH) [reply paddle/operators/matmul_op.h]: Done

**Pull request merged**